### PR TITLE
Misc zephyr native drivers fixes

### PR DESCRIPTION
**Pull request by juimonen**

### CODE REVIEW

**Review by kv2019i — APPROVED** [review]
Not sure what to make of first patch. It's mandatory to work with current Zephyr version, but whether it's cleaner than the previous version (assuming we could get the labels back), not sure anymore.

**Comment by galak** [comment]
Here's a more runtime version of `dai_get_zephyr_driver`:

```
#define GET_DEVICE_LIST(node) DEVICE_DT_GET(node),

const struct device * z_dev[] = {
        DT_FOREACH_STATUS_OKAY(intel_ssp_dai, GET_DEVICE_LIST)
        DT_FOREACH_STATUS_OKAY(intel_dmic, GET_DEVICE_LIST)
        DT_FOREACH_STATUS_OKAY(intel_alh_dai, GET_DEVICE_LIST)
};

const struct device *dai_get_zephyr_driver(uint32_t type, uint32_t index)
{
        int i;

        for (i = 0; i < ARRAY_SIZE(z_dev); i++) {
                const struct dai_config *cfg = dai_config_get(z_dev[i], DAI_DIR_BOTH);
                if ((cfg->type == type) && (cfg->dai_index == index))
                        return z_dev[i];       
        }                                      

        return NULL;
}
````

**Comment by juimonen** [comment]
> Here's a more runtime version of `dai_get_zephyr_driver`:
> 
> ```
> #define GET_DEVICE_LIST(node) DEVICE_DT_GET(node),
> 
> const struct device * z_dev[] = {
>         DT_FOREACH_STATUS_OKAY(intel_ssp_dai, GET_DEVICE_LIST)
>         DT_FOREACH_STATUS_OKAY(intel_dmic, GET_DEVICE_LIST)
>         DT_FOREACH_STATUS_OKAY(intel_alh_dai, GET_DEVICE_LIST)
> };
> 
> const struct device *dai_get_zephyr_driver(uint32_t type, uint32_t index)
> {
>         int i;
> 
>         for (i = 0; i < ARRAY_SIZE(z_dev); i++) {
>                 const struct dai_config *cfg = dai_config_get(z_dev[i], DAI_DIR_BOTH);
>                 if ((cfg->type == type) && (cfg->dai_index == index))
>                         return z_dev[i];       
>         }                                      
> 
>         return NULL;
> }
> ```

yeah thanks looks better! probably need to combine the array initialization with kconfig ifdefs... anyway will test this.

**Comment by juimonen** [comment]
so please hold with this one for a while, I got stuck with a regression in zephyr dai drivers:
https://github.com/zephyrproject-rtos/zephyr/issues/49012, so can't post the new version without testing.

File: src/lib/dai.c [hunk 1]
```diff
@@ -129,55 +129,99 @@ void dai_group_put(struct dai_group *group)
 }
 
 #if CONFIG_ZEPHYR_NATIVE_DRIVERS
+const struct device *dai_get_zephyr_driver(uint32_t type, uint32_t index)
+{
+	const struct device *drv = NULL;
```
Comment (lyakh) [inline src/lib/dai.c:134]: is this common to call a `struct device` pointer `drv`?

File: src/audio/dai-zephyr.c [hunk 2]
```diff
@@ -534,7 +534,8 @@ static int dai_playback_params(struct comp_dev *dev, uint32_t period_bytes,
 		prev = dma_block_cfg;
 		prev->next_block = ++dma_block_cfg;
 	}
-	prev->next_block = dma_cfg->head_block;
+	if (prev && prev->next_block)
```
Comment (lyakh) [inline src/audio/dai-zephyr.c:537]: is this correct? You overwrite `next_block` only if it isn't NULL? Before this patch it was set always

File: src/audio/dai-zephyr.c [hunk 3]
```diff
@@ -659,7 +660,8 @@ static int dai_capture_params(struct comp_dev *dev, uint32_t period_bytes,
...
```
Comment (lyakh) [inline src/audio/dai-zephyr.c:663]: ditto

**Review by lyakh — CHANGES_REQUESTED** [review]
just to make sure that the change is really correct, it looks suspicious

File: src/lib/dai.c (hunk 1 above)
Comment (juimonen) [reply src/lib/dai.c:134]: yeah I think in zephyr the device = driver

File: src/audio/dai-zephyr.c (hunk 2 above)
Comment (juimonen) [reply src/audio/dai-zephyr.c:537]: hmm yeah doesn't make sense... will fix

Comment (juimonen) [reply src/audio/dai-zephyr.c:537]: fixed

File: src/audio/dai-zephyr.c (hunk 3 above)
Comment (juimonen) [reply src/audio/dai-zephyr.c]: fixed

**Comment by juimonen** [comment]
new version:
depends on https://github.com/zephyrproject-rtos/zephyr/pull/49072

**Review by lgirdwood — APPROVED** [review]
Zephyr integration looks simpler now. Good stuff.

**Review by aborisovich — APPROVED** [review]

**Comment by lgirdwood** [comment]
@juimonen pls ping when the Zephyr PR is merged and then we can merge this and update our west commit (after some validation).

File: src/lib/dai.c (hunk 1 above)
Comment (lyakh) [reply src/lib/dai.c]: strange, I haven't found many such examples in Zephyr. I found indeed a couple of them, e.g. in drivers/serial/uart_lpc11u6x.c but that's about it, most use `struct device *dev` or similar. Device and driver should be pretty different things.

Comment (juimonen) [reply src/lib/dai.c]: I think we are getting into philosophical turf here... I mean "device" is the type of the struct that you instantiate to use the code associated with it. Most of the definitions are under "drivers" directory in zephyr... So maybe driver is "device + code" conceptually??? personally I think device = driver as you cant do anything without the other concept. Anyway I can change it to dev, really doesnt matter to me.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
This is apparently causing new build failure #6192

**Pull request merged**